Davi_Cruz/calculator.cpp

```diff
@@ -0,0 +1,171 @@
+#include "calculator.h"
+#include "cmath"
+
+double Calculator::primary()
+{
+    Token tk{'\0'};
+    tk = ts.get();
+    switch (tk.kind)
+    {
+    case '(':
+    {
+        double val = expression();
+        tk = ts.get();
+        if (tk.kind != ')')
+        {
+            error("primary exeption.");
+            ts.put_back(tk);
+        }
+        return val;
+    }
+    case '+':
+
+        return +primary();
+
+    case '-':
+
+        return -primary();
+
+    case func:
+    {
+        Token tk2 = ts.get();
+        if (tk2.kind != '(')
+            error("invaled declaration");
+
+        if (tk.name == sqrtkey)
+        {
+
+            ts.put_back(tk2);
+            double val = primary();
+            if (val < 0)
+                error("sqrt: square root of negative number");
+            return sqrt(val);
+        } 
+    }
+    case number:
+
+        return tk.value;
+
+    case name:
+
+        return variables.get_value(tk.name);
+
+    default:
+
+        error("primary error");
+        return 0;
+    }
+}
+double Calculator::expression()
+{
+    double left = term();
+    Token tk{'\0'};
+    while (true)
+    {
+        tk = ts.get();
+        switch (tk.kind)
+        {
+        case '+':
+            left += term();
+            break;
+        case '-':
+            left -= term();
+            break;
+
+        default:
+            ts.put_back(tk);
+            return left;
+        }
+    }
+}
+double Calculator::term()
+{
+    double left = primary();
+    Token tk{'\0'};
+    while (true)
+    {
+        tk = ts.get();
+        switch (tk.kind)
+        {
+
+        case '*':
+            left *= primary();
+            break;
+        case '/':
+        {
+            double val = primary();
+            if (val == 0)
+            {
+                error("divide by zero.");
+                left = 0;
+                break;
+            }
```

Comment:
What other construction can you use, instead of if?

Reply:
Switch